### ISSUE TITLE
Proposal: multiple formats per page using layouts

### ISSUE THREAD

**Issue by Floppy** [jekyll/jekyll#3041]
As far as I can tell, it's not possible to have two output formats for a single page/post at the moment.

We have a case where we'd like to generate an iCal file for each post, as well as the HTML version. It seems to me that a nice way to do this would be to allow multiple extensions for each layout, and generate a file for each extension in the specified layout.

For instance, if we had a post `2014-10-30-hello-world.md`, and two layouts, `_layouts/post.html` and `_layouts/post.ical`, you would end up with output in both `/2014/10/30/hello-world.html` and `/2014/10/30/hello-world.ical`.

I'll have a go at this when I find time, but I was wondering if anyone could highlight any fundamental reasons why this wouldn't work, or problems that might be encountered when writing it?

**Comment by ericmwalsh** [comment]
Is there a PR for this yet?

**Comment by kbrock** [comment]
(I withdraw my request to allow multiple html files from one document)

Anyone able to share the current proposal?

- How did it know to generate multiple pages?
- Where is the extension determined? The layout's extension or the front matter of the post?

**Comment by ericmwalsh** [comment]
Some thoughts on the proposal:

 - You should be able to choose to generate multiple pages in the config file, the layout file, and the page file (with rules of precedence applying)
 - Layouts can be listed out... ex.   layout: page-article page-article.json
 - For backward compatibility layouts listed without an extension (*.whatever) should default to HTML (or whichever user default is already specified)

**Comment by dominikwilkowski** [comment]
:+1:

**Comment by saclark** [comment]
:+1: I wrote a plugin a while back to accomplish this (https://github.com/saclark/jekyll-multipost) but standard support for this would be ten times better.

**Comment by nfreear** [comment]
:+1:  for this! I really need HTML and JSON outputs for a project I'm currently on.

**Comment by parkr** [comment]
Hey @Floppy, been thinking about this a lot and we should implement this for v3.2 or v3.3. If you have time, I was thinking we'd go this route:

1. Each page/post would define the output formats it wants **in addition to the one assigned by its source filename** (e.g. `html` for `index.md`). Front matter defaults would also apply, so doing this for a larger number of files would be trivial.
2. Those formats would use the same layout as for the HTML page, but with the new ending, e.g. `_layouts/docs.json` when `layout: docs` in the document.
3. The layout would define the behaviour. Whether you want JSON, CSV, XML, or some other output, the layout would be in charge of manipulating the data into the desired output.
4. They would have identical permalinks with the exception of the extension. In the case of `/directory/` permalinks, it would be accessible at `/directory/index.json` for consistency.
5. Accessing this link in Liquid would be `{{ page.json_url }}`, `{{ page.xml_url }}`, etc. The extension would be stripped of all non-alphanumerical characters and would be used as a prefix for `_url`.

**If you comment, please do not simply write :+1:.  If you do, your comment will be removed as it produces unnecessary notifications for others. Please react with a :+1: emoji instead, or offer a critique of the above points. Thanks!**

**Comment by Floppy** [comment]
This sounds like a very good approach @parkr, very much in line with my original thoughts. Also really pleased to see that this might go into core, because I'd dearly love to use it on github pages.

I think iCalendar format (for events, etc) is a great core use case, and people will use many more, so things like the liquid `.json_url` should be flexible to any format the user chooses, without needing to be predefined.

**Comment by parkr** [comment]
@Floppy Cool, I'm glad you like the proposal. To be clear, if you define a `.ical` output format for your page and you define the appropriate `.ical` layout, you're golden. We're going to have to stick to text-only formats for now. The `Jekyll::Generator` plugin type should be used for more complicated outputs.

**Comment by parkr** [comment]
@jekyll/ecosystem @jekyll/build – would anyone like to take this on for v3.2? Outlined in https://github.com/jekyll/jekyll/issues/3041#issuecomment-196987588

**Comment by dominikwilkowski** [comment]
Just to clarify: this would mean you can have multiple layouts only for outputs that have different formats. Or in other words: You can't output two HTML files for one md file.

We have a use-case where we need several HTML pages rendered differently just through different layouts from the same md dataset. I guess you could argue this is edge-case? Anyone else have that use-case? (vote with :+1: emoji instead of comments)

**Comment by Floppy** [comment]
Should there be an option to specify the `content-type` for a layout/extension, or would we limit folks to the stuff already in the `mime-types` gem? I suspect the latter is fine, as it;s pretty comprehensive, but I thought I'd ask :)

And iCalendar is `.ics` not `.ical` of course, my mistake in the original proposal there.

**Comment by pathawks** [comment]
> You can't output two HTML files for one md file.

What if I want a standard HTML and an AMP HTML file for each post?

**Comment by spudowiar** [comment]
You could always just make `layout` an Array and a String (e.g. you could have `layout: post` or `layout: [post, metadata.json]`)

**Comment by envygeeks** [comment]
That's a blocker.  Jekyll already has performance problems. Now we want to create more performance problems before addressing the old performance problems?

**Comment by spudowiar** [comment]
@envygeeks It should make no difference to performance if you only use one layout (prepare suffer if you decide to use multiple :P)

**Comment by envygeeks** [comment]
@spudowiar it's going to add to the times no matter what.  While trivial over a large site it will add up, in the case of our sites where we have 20K some-odd pages (on some sites - or 3K is the average) this is unacceptable until other problems are addressed.  And as Jekyll moves from version to version we keep seeing more and more build time.

**Comment by Floppy** [comment]
@dominikwilkowski @pathawks in your examples, how would the URLs differ for the different outputs? That might point towards a way it might work.

BTW, @pathawks by AMP I'm assuming you mean [Accelerated Mobile Pages](https://www.ampproject.org/)?

**Comment by spudowiar** [comment]
@envygeeks Having this feature at all would add the same amount of overhead, though?

**Comment by saclark** [comment]
@Floppy What about something similar to [how my plugin handles it](https://github.com/saclark/jekyll-multipost#usage)?

**Comment by envygeeks** [comment]
@spudowiar nobody said I supported this feature at all ;) I just targeted the overhead that added to the overhead everybody already agreed to accept.

**Comment by envygeeks** [comment]
@spudowiar but I will hold my official opinion until I see code somebody produces.

**Comment by spudowiar** [comment]
@envygeeks Ah, I misunderstood you, I thought you were against my method of implementation. Mea culpa :)

**Comment by dominikwilkowski** [comment]
@Floppy I can think of two ways to deal with this.

### version 1
The layout must have different names as layouts are all in the same folder so attaching the layout name to the url if there are multiple layouts specified.

### version 2
When multiple layouts are selected in front matter you also have to name the urls e.g.:
```
---
layout:
  layout1
    url1
  layout2
    url2
---
```

I do like version 1 and how @saclark [plugin](https://github.com/saclark/jekyll-multipost#usage) handles it.

**Comment by Floppy** [comment]
I've got a very early work in progress in #4742. At the moment most of the above design discussion isn't reflected in there, I'm just exploring for now :)

**Comment by jekyllbot** [comment]
This issue has been automatically marked as stale because it has not been commented on for at least
three months.

The resources of the Jekyll team are limited, and so we are asking for your help.

If you can still reproduce this error on the <code>3.1-stable</code> or <code>master</code> branch,
please reply with all of the information you have about it in order to keep the issue open.

If this is a feature request, please consider building it first as a plugin. Jekyll 3 introduced
[hooks](http://jekyllrb.com/docs/plugins/#hooks) which provide convenient access points throughout
the Jekyll build pipeline whereby most needs can be fulfilled. If this is something that cannot be
built as a plugin, then please provide more information about why in order to keep this issue open.

Thank you for all your contributions.

**Comment by dominikwilkowski** [comment]
too bad :(

**Comment by parkr** [comment]
Closed due to staleness. I'll re-open.

**Comment by dominikwilkowski** [comment]
Just out of interest, how is this proposal going?
@Floppy was working on a PR at some point?

**Comment by Floppy** [comment]
I was, #4742, but I didn't get very far before having to deal with other things. It's still on my one-day list, but I don't get a lot of time these days, it seems.

**Comment by dnl** [comment]
To add an use case: I'd kind of like a way to output the raw markdown as a second file (raw-ish anyway,  removing front-matter and outputting page title, date into the document).

Something similar to putting `.text` on the end of daringfireball articles:  http://daringfireball.net/2016/08/apple_gun_emoji.text

**Comment by shawnbot** [comment]
Huge +1 from me on this, too. Among many other things, this would make it possible to make style guides (a la [Fractal] or [Pattern Lab]) with Jekyll, by allowing each item in a component collection to output both a "framed" page and an "unframed" one that could be placed in an `<iframe>` by the former.

[Fractal]: http://fractal.build/
[Pattern Lab]: http://patternlab.io/

**Comment by goodevilgenius** [comment]
Another great use case for this would be to have regular html layout, plus [AMP layout](https://www.ampproject.org/) for the same posts.

**Comment by ondovb** [comment]
Hack that worked for me - just use symlinks to make multiple collections from the same source files. Github Pages seems to handle it fine.

eg, in project dir:

    >ln -s _collection1/ _collection2/

in `_config.yml`:

    collections:
      collection1:
        output: true
      collection2:
        output: true
     
    defaults:
     -
       scope:
         path: "_collection1"
       values:
         layout: layout1
     -
       scope:
         path: "_collection2"
       values:
         layout: layout2

**Comment by tf2manu994** [comment]
I would also love this feature for AMP pages

**Comment by jekyllbot** [comment]
This issue has been automatically marked as stale because it has not been commented on for at least two months.

The resources of the Jekyll team are limited, and so we are asking for your help.

If this is a **bug** and you can still reproduce this error on the <code>3.3-stable</code> or <code>master</code> branch, please reply with all of the information you have about it in order to keep the issue open.

If this is a **feature request**, please consider building it first as a plugin. Jekyll 3 introduced [hooks](http://jekyllrb.com/docs/plugins/#hooks) which provide convenient access points throughout the Jekyll build pipeline whereby most needs can be fulfilled. If this is something that cannot be built as a plugin, then please provide more information about why in order to keep this issue open.

This issue will automatically be closed in two months if no further activity occurs. Thank you for all your contributions.

**Comment by pathawks** [comment]
If this feature were available in core, I would immediately create a plugin that used it to create Amp versions of pages.

I don't have time to build this feature myself right now, but I would very much like to see it exist.

**Comment by jekyllbot** [comment]
This issue has been automatically marked as stale because it has not been commented on for at least two months.

The resources of the Jekyll team are limited, and so we are asking for your help.

If this is a **bug** and you can still reproduce this error on the <code>3.3-stable</code> or <code>master</code> branch, please reply with all of the information you have about it in order to keep the issue open.

If this is a **feature request**, please consider building it first as a plugin. Jekyll 3 introduced [hooks](http://jekyllrb.com/docs/plugins/#hooks) which provide convenient access points throughout the Jekyll build pipeline whereby most needs can be fulfilled. If this is something that cannot be built as a plugin, then please provide more information about why in order to keep this issue open.

This issue will automatically be closed in two months if no further activity occurs. Thank you for all your contributions.

**Comment by pup** [comment]
@jekyllbot not stale

**Comment by pathawks** [comment]
> If this is a feature request, please consider building it first as a plugin.

This might be the way to go. Thinking about this, there are all sorts of weird problems that would arise.

Which format would the `{% link %}` tag point to? If a URL is specified, which format gets that URL and what happens to other formats?

I feel like this is a pretty big structural change to Jekyll, not in code but in expectations. This would almost certainly need a major version bump.

@pnn I worry that pinning this makes it seem like somebody is working on this, and users can expect this feature soon.

**Comment by pup** [comment]
@pathawks do you expect this issue to be open for much longer though? cuz if it does, we're gonna have to deal with jekyllbot every two months

**Comment by pathawks** [comment]
I'm suggesting maybe this issue should be closed, and somebody could develop a plugin if they needed this functionality.

**Comment by ashmaroli** [comment]
There's already a plugin for this out there..( [comment](https://github.com/jekyll/jekyll/issues/3041#issuecomment-94605383) by @saclark ).. hopefully it could be polished, updated and ported to Core?

**Comment by timwis** [comment]
Yeah part of core makes a big difference because of gh-pages support. For JKAN, for example.

**Comment by JasonYao** [comment]
Just wanted to add to @ondovb's answer above that it works, but with a notable requirement: the 
symlinked collection has to be listed **before** the real one, else it throws an error when restarting with `bundle exec jekyll serve`.

The error that you'd see would be something along the lines of:
```
jekyll 3.4.3 | Error:  (PATH_TO_PROJECT_ROOT/_config.yml): did not find expected key while parsing a block mapping at line 6 column 1
```

In my use case, I wanted to generate two sets of HTML from the same source file without duplication, one an HTML partial that was just a summary, and the other the full HTML.

## Working live code
- `ln -s projects projects_full`
- Inside `_config.yml`:
```yml
collections:
  projects:
    permalink: /projects/partial/:path/
    output: true
  projects_full:
    permalink: /projects/:path/
    output: true

defaults:
  -
    scope:
      path: "_projects_full"
    values:
      layout: project_full
# NOTE: The symlinked directory has to be above the normal one for fuckery reasons.
  -
    scope:
      path: "_projects"
    values:
      layout: project
```

This worked for me, and generated the correct output.

**Comment by grokcode** [comment]
I'd love to see this feature in core. It would make it possible to have blog comments on their own page without resorting to plugins or JavaScript hackery.

**Comment by adamvoss** [comment]
@timwis GitHub Pages support can be achieved without a feature being merged into Jekyll core if you can get it added to the [white-listed plugins](https://github.com/github/pages-gem).

**Comment by timwis** [comment]
@adamvoss any idea how that process works? Seems very selective.

**Comment by adamvoss** [comment]
@timwis That is a question you would need to raise over there.  The only thing I know for certain is from https://github.com/github/pages-gem/issues/355#issuecomment-287767649 indicating included plugins must be under test:

> we wouldn't be able to consider any plugin without unit and integration tests for the desired behavior.

**Comment by parkr** [comment]
I remember feeling very overwhelmed by https://github.com/jekyll/jekyll/issues/4742 ... maybe I need to write this feature myself. That's what ended up happening with Collections back in the day. If we can come up with some tests in `features/` which define how a user would tell Jekyll that it wants multiple formats, I could take it from there.

**Comment by Floppy** [comment]
I never got very far with #4742... sorry about that!

**Comment by parkr** [comment]
@Floppy No problem at all! I just don't know how best to integrate this into our flow. It's not super often that a markdown page wants to be output in 2 different formats on a per-page basis, so I don't really have a clear set of use-cases that can help me evaluate how this feature should work!

**Comment by DirtyF** [comment]
@parkr FWIW, Hugo supports [custom output formats](https://gohugo.io/templates/output-formats/) for media types.

**Comment by parkr** [comment]
@DirtyF Ah, neat! Looks like they take the YAML approach as well:

```text
---
date: "2016-03-19"
outputs:
- html
- amp
- json
---
```

**Comment by DirtyF** [comment]
@parkr yes either in a page's Front Matter or in site config.

> Which format would the {% link %} tag point to? If a URL is specified, which format gets that URL and what happens to other formats?

@pathawks Good question. In Hugo each page has [output variables](https://gohugo.io/templates/output-formats/#link-to-output-formats) that can be used to link to different kind of outputs.

**Comment by ashmaroli** [comment]
Awesome insights @DirtyF If its possible in go-lang, it can be replicated in Ruby too, right? (barring the performance issues from iterating a Ruby Collection)

**Comment by DirtyF** [comment]
@ashmaroli yeah I heard @envygeeks concerns about build time and maybe this should only be addressed once #6252 is done, so that we can maybe reconsider #5344 ?

**Comment by ashmaroli** [comment]
Likely, so.. Any idea if we could benchmark PRs via CI? i.e. Have it checked loosely if the current patch has any effect on the performance of the overall library?

**Comment by envygeeks** [comment]
You can add a custom script to run a check of the code before the pull, and a check of the pull and to run in each and compare, if it lasts more than a second though, it could be bad, and wrong, considering there is many factors to consider on Travis-CI (steal time, io wait, etc...)

**Comment by LyndisCealin** [comment]
as @saclark mentioned, he has written this plugin: https://github.com/saclark/jekyll-multipost

So I changed the permalinkBuilder to this:

```ruby
module PermalinkBuilder
    extend self

    def get_adjusted_permalink(resource, layout)
      layout_path = CGI.escape(layout)
      url = resource.url
      ext = File.extname(url)

      if url.include?(':layout')
        return url.gsub(/:layout/, layout_path)
      end

      if ext.empty?
        "#{url}/#{layout_path}/"
      else if layout_path.include?('story')
            url.gsub(/\/$|#{ext}$/) { |url_end| "#{url_end}"}
           else if layout_path.include?('blog_index')
            url.gsub(/\/$|#{ext}$/) { |url_end| "#{url_end}"}
           else if layout_path.include?('rss')
            url.sub('.html', '.xml')
           else
            url.gsub(/\/$|#{ext}$/) { |url_end| "-#{layout_path}#{url_end}" }
           end 
          end
        end
      end
    end
  end
```

Please ignore the 'story' and 'blog_index' part. Important is the 'rss' option.
Every layout rendered with the layout 'rss' has now a '.xml' extension.
So I think with this plugin, you can do everything mentioned in this thread before.

**Comment by jekyllbot** [comment]
This issue has been automatically marked as stale because it has not been commented on for at least two months.

The resources of the Jekyll team are limited, and so we are asking for your help.

If this is a **bug** and you can still reproduce this error on the <code>3.3-stable</code> or <code>master</code> branch, please reply with all of the information you have about it in order to keep the issue open.

If this is a **feature request**, please consider building it first as a plugin. Jekyll 3 introduced [hooks](http://jekyllrb.com/docs/plugins/#hooks) which provide convenient access points throughout the Jekyll build pipeline whereby most needs can be fulfilled. If this is something that cannot be built as a plugin, then please provide more information about why in order to keep this issue open.

This issue will automatically be closed in two months if no further activity occurs. Thank you for all your contributions.

**Comment by jekyllbot** [comment]
This issue has been automatically marked as stale because it has not been commented on for at least two months.

The resources of the Jekyll team are limited, and so we are asking for your help.

If this is a **bug** and you can still reproduce this error on the <code>3.3-stable</code> or <code>master</code> branch, please reply with all of the information you have about it in order to keep the issue open.

If this is a **feature request**, please consider building it first as a plugin. Jekyll 3 introduced [hooks](http://jekyllrb.com/docs/plugins/#hooks) which provide convenient access points throughout the Jekyll build pipeline whereby most needs can be fulfilled. If this is something that cannot be built as a plugin, then please provide more information about why in order to keep this issue open.

This issue will automatically be closed in two months if no further activity occurs. Thank you for all your contributions.

**Comment by pup** [comment]
not quite stale yet, i don't want to give up on this yet

**Comment by Convincible** [comment]
Just chiming in that multiple output formats seems an obvious missing feature which I'd love to use, particularly around making site data more easily consumable by other systems. Currently thinking about how I can plugin this, presumably as a Generator, but would love this built in.

Functionality I'd be after:

* `layout: [a, b, c]` syntax to me seems simplest – we simply extend from being able to define 1 to being able to define many
* Layouts are extension agnostic: could be a.html, b.html, c.json – doesn't matter. They are referred to by their extensionless filename (as per existing functionality). Up to the developer to avoid name clashes.
* These could in turn be nested inside other, multiple layouts. That is, a.html could specify: `layout: [d, e]`. These again are agnostic e.g. d.amp.html, e.ics
* The final file at the end takes the extension of the highest-level layout.
* In cases of filename clash, layouts defined later append the layout name to the filename, e.g. page_b.ics (whereas the first defined layout would be e.g. page.html)

Why this way?

1. No reason to limit this to the use case of multiple unique formats. What about multiple different outputs in the same format?
2. Nesting makes sense. E.g. you might build up a JavaScript fragment by nesting JS layouts, which is then ultimately pumped into an HTML layout's head. That is, formats often contain other formats, but the biggest container defines the overall format.

**Comment by kristoffernolgren** [comment]
I need would need this aswell. I used the symlink version first, but my hostingprovider (cloudcannon) won't support it. In my case I need to render a collection with a special template for links from insied an app. that uses embeded web-views. So I need want to use the same source markdown and different templates/settings.

**Comment by IssuehuntBot** [comment]
@0maxxam0 has funded $60.00 to this issue. [See it on IssueHunt](https://issuehunt.io/repos/65252/issues/3041)

**Comment by Floppy** [comment]
Wow, 4 years old and I've still not started working on this. It'll be old enough to go to school soon. 🍰

**Comment by kbrock** [comment]
@Floppy the multipost code looks like a  good start.
Also looks like they have an answer for multiple of the same format (html) and conflicting layouts

**Comment by IssuehuntBot** [comment]
@issuehuntfest has funded $100.00 to this issue. [See it on IssueHunt](https://issuehunt.io/repos/65252/issues/3041)

**Comment by IssuehuntBot** [comment]
@dirtyf has cancelled their pull request. [See it on IssueHunt](https://issuehunt.io/repos/65252/issues/3041)

**Comment by IssuehuntBot** [comment]
@rororofff has funded $2.00 to this issue.

---
- Submit pull request via [IssueHunt](https://issuehunt.io/repos/65252/issues/3041) to receive this reward.
- Want to contribute? Chip in to this issue via [IssueHunt](https://issuehunt.io/repos/65252/issues/3041).
- Checkout the [IssueHunt Issue Explorer](https://issuehunt.io/issues) to see more funded issues.
- Need help from developers? [Add your repository](https://issuehunt.io/users/47737115/repositories) on IssueHunt to raise funds.

**Comment by IssuehuntBot** [comment]
@0maxxam0 has funded $40.00 to this issue.

---
- Submit pull request via [IssueHunt](https://issuehunt.io/repos/65252/issues/3041) to receive this reward.
- Want to contribute? Chip in to this issue via [IssueHunt](https://issuehunt.io/repos/65252/issues/3041).
- Checkout the [IssueHunt Issue Explorer](https://issuehunt.io/issues) to see more funded issues.
- Need help from developers? [Add your repository](https://issuehunt.io/users/40680108/repositories) on IssueHunt to raise funds.

**Comment by IssuehuntBot** [comment]
@serima has funded $2.00 to this issue.

---
- Submit pull request via [IssueHunt](https://issuehunt.io/repos/65252/issues/3041) to receive this reward.
- Want to contribute? Chip in to this issue via [IssueHunt](https://issuehunt.io/repos/65252/issues/3041).
- Checkout the [IssueHunt Issue Explorer](https://issuehunt.io/issues) to see more funded issues.
- Need help from developers? [Add your repository](https://issuehunt.io/users/1130921/repositories) on IssueHunt to raise funds.

**Comment by IssuehuntBot** [comment]
@amthe has funded $2.00 to this issue.

---
- Submit pull request via [IssueHunt](https://issuehunt.io/repos/65252/issues/3041) to receive this reward.
- Want to contribute? Chip in to this issue via [IssueHunt](https://issuehunt.io/repos/65252/issues/3041).
- Checkout the [IssueHunt Issue Explorer](https://issuehunt.io/issues) to see more funded issues.
- Need help from developers? [Add your repository](https://issuehunt.io/r/new) on IssueHunt to raise funds.

**Comment by chetabahana** [comment]
I would like to share an alternative way using a plug in called [jekyll-redirect-from](https://github.com/jekyll/jekyll-redirect-from) that also available on [github pages](https://pages.github.com/versions/) by default.

Based on your case of a post called `2014-10-30-hello-world.md`, and two layouts, `_layouts/post.html` and `_layouts/post.ical` with an expected different output in both `/2014/10/30/hello-world.html` and `/2014/10/30/hello-world.ical`.

Using different extension but with the same name of layout will not be possible, so we will need to move them in the a different folder than layout.  In this case `_includes` folder will be used.

In `_includes/post.html`
```
move the contents of your _layouts/post.html here..
```
In `_includes/post.ical`
```
move the contents of your _layouts/post.ical here..
```
Arrange redirection in your post  `2014-10-30-hello-world.md`
```
---
layout: my_layout
redirect_from:
  - /2014/10/30/hello-world.html
  - /2014/10/30/hello-world.ical
---
Hello World
```
In `_layouts/my_layout.html` use output to render the content
```
{{ page.output }}
```

 In `_layouts/redirect.html`
```
{% include {{ page.redirect.from | split: "." | last | prepend: "post." }} %}
```
As result the output of `/2014/10/30/hello-world.html` and `/2014/10/30/hello-world.ical` will take `__includes/post.html` and `__includes/post.ical` respectively.

**Comment by Floppy** [comment]
Oh, @chetabahana, that is a clever hack - I love it :)

**Comment by katcaola** [comment]
A bit late to the party, but I would also like to see this feature implemented.

I would like to have multiple layouts available per post so that I can branch what a post looks like based on how you got to it.

Example: All posts list -> post -> post list still show all posts while viewing post. Simultaneously: Filtered posts list -> post -> post list still show filtered posts while viewing post

![wireFrame](https://user-images.githubusercontent.com/5624265/75115185-3a014b80-562a-11ea-90cb-06994edaae62.png)

**Comment by bloombar** [comment]
Almost a decade later, the generalized plural option for layouts, where either `layout: first` or `layouts: [first, second]` seems like it would still work and would be valuable and necessary.  The complication from a user's perspective would be the conflict of permalinks - if a single source document, say `foo.md` (which would usually have a single permalink such as `/foo`) specifies the use of two or more layouts, how are permalinks handled to avoid conflict?  

I propose the following:
- if only one layout, e.g. `layout: first`, use the usual permalink, e.g. `/foo`
- if two or more layouts, e.g. `layouts: [first, second]`,  the first layout in the list uses the usual permalink, e.g. `/foo` for first layout, while the subsequent layouts have the layout name suffix on the permalink, e.g. `/foo-second`

**Comment by ashmaroli** [comment]
Wow! This issue has been incubating for almost a decade now.. Wonder how many of the those desired this feature is still using Jekyll...

**Comment by elisehein** [comment]
> Wow! This issue has been incubating for almost a decade now.. Wonder how many of the those desired this feature is still using Jekyll...

🖐️

**Comment by FrankKooij** [comment]
> Wow! This issue has been incubating for almost a decade now.. Wonder how many of the those desired this feature is still using Jekyll...

We are using Jekyll for NSDI data specifications and need this feature for content negotiation.

**Comment by Convincible** [comment]
Still using Jekyll. Still desire this feature!

Agree with @bloombar's approach which should apply only in the case of a filename clash. E.g.

Files in _layouts:
* a.html
* b.amp
* c.json
* d.html

Inside c.json is also the frontmatter: ```layout: d``` (i.e. this layout nests itself in another)

So:

| Input file | Layouts | Output files | Note |
| --- | --- | --- | --- |
| page.md | `a, b` | page.html, page.amp | |
| page.md | `a, d` | page.html, page_d.html | Filename clash |
| page.md | `a, c` | page.html, page_d.html | Filename clash |

**Comment by bloombar** [comment]
@Convincible - thanks for the detailed description.  There are two issues here that might be best kept separate:
1. allowing multiple layouts (regardless of file extension)
2. supporting layouts with other content types and file extensions besides just `text/html` and `.html`.

The first seems relatively straightforward, if a simple rule is devised like the one you and I have suggested.  I believe the second issue might be a bit more challenging, since jekyll does not seem to currently support any content type besides `text/html` (correct me if that's just my own machine) and I suspect from briefly wrangling with the `jeyll-redirect-from` plugin (which has problems related to non-HTML content types and file extensions in the [redirect-to](https://github.com/jekyll/jekyll-redirect-from/issues/252#event-7582397202) or [redirect-from](https://github.com/jekyll/jekyll-redirect-from/issues/252#event-7582397202) directives) that the problem is not with jekyll itself, but rather with whichever web server is used to serve the Jekyll pages, webrick or other.

**Comment by cycomachead** [comment]
Put me in the camp of someone who also wants this, and who has used Jekyll (off and on) for much of this time. :) 

What have other folks been doing in the meantime to get multiple outputs?

I might have a slightly unexpected use case for a project soon, which is to generate both a PDF and HTML version of page.
In that sense, a format of sorts is probably more useful to me than just text layouts, since it could more clearly hook into a converter. 

However, it seems like [Hooks](https://jekyllrb.com/docs/plugins/hooks/) should be able to do this, perhaps `:pre_render` or `:post_render` depending on which format it's easier to write the secondary output from?

**Comment by seshrs** [comment]
> I might have a slightly unexpected use case for a project soon, which is to generate both a PDF and HTML version of page.

I have a similar use case, but rather than create a Jekyll plugin, I personally found it easier to use my own custom GitHub Actions workflow to render the PDFs _after_ running Jekyll. 

https://github.com/seshrs/build-primer-spec-action/blob/62703e09107f7d70f6e83f2a60b2690bbe33d98d/action.yml#L77,L92

**Comment by seroperson** [comment]
@mattr- @ashmaroli I want to give it a try. I see there was quite a big discussion, but seems like everybody agree with things @Convincible have described: be able to define several layouts like `layout: [...]`, nesting, filename clashing and so on. I can start playing around it.
What do you think? Maybe you can highlight some must-have things to be implemented?

**Comment by bloombar** [comment]
@seroperson - 

We would also need to avoid [permalink](https://jekyllrb.com/docs/permalinks/) clashes, which might be more common than filename clashes if `pretty` permalinks are configured. A few ideas:

## Option 1

Extending @Convincible's [pattern above](https://github.com/jekyll/jekyll/issues/3041#issuecomment-1387190290), the following logic would append the layout name as a suffix to the filename AND the permalink for all but the first layout, e.g.

| Input file | Layouts | Output files           | Pretty Permalink | Note                            |
| ---------- | ------- | ---------------------- | ---------------- | ------------------------------- |
| page.md    | `a, b`  | page.html, page.amp    | /page, /page_b   | Permalink clash                 |
| page.md    | `a, d`  | page.html, page_d.html | /page, /page_d   | Permalink clash, Filename clash |
| page.md    | `a, c`  | page.html, page_d.html | /page, /page_d   | Permalink clash, Filename clash |

## Option 2

Alternatively, we could only use pretty permalinks only for the first layout and provide only filename access to subsequent layouts, e.g.:

| Input file | Layouts | Output files           | Pretty Permalink    | Note                            |
| ---------- | ------- | ---------------------- | ------------------- | ------------------------------- |
| page.md    | `a, b`  | page.html, page.amp    | /page, /page.amp    |   Permalink clash                              |
| page.md    | `a, d`  | page.html, page_d.html | /page, /page_d.html | Permalink clash, Filename clash |
| page.md    | `a, c`  | page.html, page_d.html | /page, /page_d.html | Permalink clash, Filename clash |

My personal preference is Option 2, since it keeps the permalinks closest to the filenames.  But either is an improvement over the current state.

**Comment by mattr-** [comment]
I'd actually prefer an initial implementation closer to the original requests and some of the original attempted designs such as #4742

**Comment by alexey-pelykh** [comment]
This issue will soon go to college :)

**Comment by pikesley** [comment]
> This issue will soon go to college :)

It can't be bargained with, it can't be reasoned with, it doesn't feel pity or remorse or fear, and it absolutely will not stop, ever
